apps/emqx_bridge_redis/src/emqx_bridge_redis_connector.erl

```diff
@@ -162,6 +169,11 @@ on_batch_query(
             Error
     end.
 
+trace_format_commands(Commands0) ->
+    Commands1 = [[unicode:characters_to_list(S) || S <- C] || C <- Commands0],
+    Commands2 = [lists:join(" ", C) || C <- Commands1],
+    unicode:characters_to_binary(lists:join("; ", Commands2)).
```

Comment:
`unicode:characthers_to_binary` works with `iodata()` type, so there is no need to convert them into a string list then join with space then convert back to binary.

Reply:
Sorry, yes, you are right. The conversion on line 173 was done as I first used string:join which only worked with list strings but then I figured out that I could just as well use lists:join but forgot to remove the conversion on the previous line. While looking at another issue I found that it is actually best_effort_json that cause the problem we are trying to solve here. best_effort_json has heuristics that tries to convert things that looks like IO data into into strings. But this heuristics in at least two places (this connector and the oracle connector). I think perhaps a better way to solve it is to wrap the list in something that forces best_effort_json to treat it as a list but then we have to handle this wrapping in all formatters...

Reply:
if you mean this macro?
`-define(IS_STRING(String), (is_list(String) orelse is_binary(String))).`

yes, the macro can be optimized to check if the list style string also meets this condition: `String =/=[] andalso is_integer(hd(String))`

However, I actually mean this:
```
    Commands1 = [lists:join(" ", C) || C <- Commands0],
    unicode:characters_to_binary(lists:join("; ", Commands1)).
```

Reply:
Should we use unicode tranformations at all to dump commands? 🤔 
Redis is not unicode aware, it operates with byte sequences as commands and operands.
Doesn't converting to unicode bring umbiguity of what actually was sent to redis 🤔

Reply:
redis is not unicode aware, then it's EMQX's responsibility to make sure unicode code-points are turned into byte streams (utf8), so I think it does make sense to format logs as utf8 chars

our config is read as utf8, and the templates are rendered utf8